src/core/Track.cpp

```diff
@@ -98,35 +89,36 @@ Track::~Track()
  *  \param tt The type of track to create
  *  \param tc The track container to attach to
  */
-Track * Track::create( Type tt, TrackContainer * tc )
+Track* Track::create(Type type, TrackContainer* trackContainer)
 {
-	Engine::audioEngine()->requestChangeInModel();
-
-	Track * t = nullptr;
-
-	switch( tt )
+	auto track = static_cast<Track*>(nullptr);
+	switch (type)
 	{
-		case Type::Instrument: t = new class InstrumentTrack( tc ); break;
-		case Type::Pattern: t = new class PatternTrack( tc ); break;
-		case Type::Sample: t = new class SampleTrack( tc ); break;
-//		case Type::Event:
-//		case Type::Video:
-		case Type::Automation: t = new class AutomationTrack( tc ); break;
-		case Type::HiddenAutomation:
-						t = new class AutomationTrack( tc, true ); break;
-		default: break;
+	case Type::Instrument:
+		track = new InstrumentTrack(trackContainer);
+		break;
+	case Type::Pattern:
+		track = new PatternTrack(trackContainer);
+		break;
+	case Type::Sample:
+		track = new SampleTrack(trackContainer);
+		break;
+	case Type::Automation:
+	case Type::HiddenAutomation:
+		track = new AutomationTrack(trackContainer, type == Type::HiddenAutomation);
+		break;
+	default:
+		break;
 	}
 
-	if (tc == Engine::patternStore() && t)
+	assert(track && "Track::create failed");
+
 	{
```

Comment:
The { was for an if block right? Since you replaced that with an assert, shouldn't that block go?

Reply:
The braces is because of RAII. `Engine::audioEngine::requestChangesGuard` returns a object that calls `AudioEngine::requestChangeInModel` on construction and `AudioEngine::doneChangeInModel` on destruction. The braces specify the lifetime of the guard.

However, it looks like I don't need the braces because of where I placed the guard (right when we add the track and that is it). I think what happened was before there was still some code that needed to be ran after we have added the track, but I moved that code out of this function and forgot to remove them.

Reply:
Also, since you use requestChangesGuard, shouldn't you also use doneChangesGuard.?

Reply:
`requestChangesGuard` is an RAII wrapper over `requestChangeInModel` and `doneChangeInModel`. Notice the distinction in name. `requestChangesGuard` basically calls `requestChangeInModel` when the object is constructed, and `doneChangeInModel` when it is being destroyed.

Reply:
Ahh mb.

Reply:
If you are wondering why I'm not simply using `requestChangeInModel` and `doneChangeInModel` here, it's because RAII is the preferred approach when it comes to things like this. This is similar to how `std::lock_guard` works, which is simpler than calling `std::mutex::lock` and `std::mutex::unlock` and avoids the problem that could happen when you forget to call `std::mutex::unlock`.

The codebase hasn't fully made the transition yet. The whole locking mechanism will need to be removed anyways.